module/spl/spl-kmem-cache.c

```diff
@@ -202,8 +202,22 @@ kv_alloc(spl_kmem_cache_t *skc, int size, int flags)
 	if (skc->skc_flags & KMC_KMEM) {
 		ASSERT(ISP2(size));
 		ptr = (void *)__get_free_pages(lflags, get_order(size));
+	} else if (skc->skc_flags & KMC_KVMEM) {
+		ptr = spl_kvmalloc(size, lflags);
 	} else {
-		ptr = __vmalloc(size, lflags | __GFP_HIGHMEM, PAGE_KERNEL);
+	/*
+	 * GFP_KERNEL allocations can safe use kvmalloc allocator which
+	 * may increase performance since vmalloc overhead might not
+	 * be applicated since enough contiguous memory was available
+	 * at allocation time. unfortunatly the slab statistic will
+	 * show allocations as VMEM without any care if kmalloc was
+	 * used in reality. but this is just cosmetic
+	 */
```

Comment:
nit: might be GitHub's formatting by the indention here doesn't look right.  A few typo's in the comment too.  How about:

```
         * GFP_KERNEL allocations can safely use kvmalloc which may
	 * improve performance by avoiding vmalloc lock overhead when
	 * sufficient contiguous memory is available.  The slab statistics
	 * will incorrectly report this as a vmem allocation, but that is
         * is purely cosmetic.
```

Reply:
mmh lock overhead. i dont know what you mean with lock overhead. the overhead i see is different. vmalloc normally does not allocated anything, but on first access of memory which causes a high latency spike on first access of that memory area. second: virtual memory is always slightly slower than contiguous physical memory due the memory mapped mmu handling of memory. especially since the tlb space is limited (the size of the tlb is cpu specific of course)
so depending on the memory conditions it can turn even very slow which i want to avoid for specific memory areas like decompression workspaces if it can be avoided. such workspaces should not be swapped in worst case if it was preallocated successfully

Reply:
Well, what about this one?

~~~
* GFP_KERNEL allocations can safely use kvmalloc which may improve
* performance by avoiding a) high latency caused by vmalloc's
* on-access allocation, b) performance loss due to MMU memory address
* mapping and c) vmalloc locking overhead.
* This has the side-effect that the slab statistics will incorrectly
* report this as a vmem allocation, but that is is purely cosmetic.
* To avoid serious performance drop due to swapping kvmalloc will
* onlybe used when a) sufficient contiguous memory is available and
* b) size <= spl_kmem_alloc_max.
~~~

Reply:
That's much better, with I believe one small correction.  On Linux kernel memory is always pinned and cannot be swapped, so we should drop that last paragraph.  Unless I misunderstood the point @BrainSlayer is making, in which case let's be clearer.

Reply:
I dropped the swapping part:

~~~
...
		 * However, kvmalloc will only be used when a) sufficient contiguous
		 * memory is available and b) the allocated memory is less than
		 * spl_kmem_alloc_max.
~~~

Reply:
Hmmm.... don't we need a check for size <= spl_kmem_alloc_max in `kv_alloc`, too?